x/wasm/keeper/keeper.go

```diff
@@ -1005,6 +1029,116 @@ func (k Keeper) newQueryHandler(ctx sdk.Context, contractAddress sdk.AccAddress)
 	return NewQueryHandler(ctx, k.wasmVMQueryHandler, contractAddress, k.gasRegister)
 }
 
+func (k Keeper) updateSizeForContract(ctx sdk.Context, contractAddress sdk.AccAddress, sizeChange int64) error {
+	// always charge rent first before updating size, as the charging logic assumes constant size since last
+	// charged block
+	if err := k.chargeRent(ctx, contractAddress); err != nil {
+		return err
+	}
+	sizeKey := types.GetContractStoreSizePrefix(contractAddress)
+	store := ctx.KVStore(k.storeKey)
+	currentSize, err := k.getCurrentSizeForContract(ctx, contractAddress)
+	if err != nil {
+		return types.ErrUpdateContractSize.Wrap(err.Error())
+	}
+	currentSizeSigned := int64(currentSize) + sizeChange
+	if currentSizeSigned < 0 {
+		// this should not happen
+		panic("negative size for contract")
+	}
+	newBz := make([]byte, 8)
+	binary.BigEndian.PutUint64(newBz, uint64(currentSizeSigned))
+	store.Set(sizeKey, newBz)
+	return nil
+}
+
+func (k Keeper) getCurrentSizeForContract(ctx sdk.Context, contractAddress sdk.AccAddress) (uint64, error) {
+	return GetSize(ctx, contractAddress, k.storeKey)
+}
+
+func (k Keeper) chargeRent(ctx sdk.Context, contractAddress sdk.AccAddress) error {
+	rentInfo, err := k.getRentInfo(ctx, contractAddress)
+	if err != nil {
+		return err
+	}
+	duration := uint64(ctx.BlockHeight()) - rentInfo.LastChargedBlock
+	if duration == 0 {
+		// shortcut in the case of same block changes
+		k.setRentInfo(ctx, contractAddress, rentInfo)
+		return nil
+	}
+	size, err := k.getCurrentSizeForContract(ctx, contractAddress)
+	if err != nil {
+		return err
+	}
+	if size == 0 {
+		rentInfo.LastChargedBlock = uint64(ctx.BlockHeight())
+		k.setRentInfo(ctx, contractAddress, rentInfo)
+		return nil
+	}
+	rentPrice := k.GetParams(ctx).UnitRentPrice
```

Comment:
in the case of a param change for the unit rent price on block 5 (and then this contract is on block 10), are we going to assess all 10 blocks with the new param value? or would we want to assess rent against contracts for the previous param value at the time of updating the param? I'm fine with either approach, just implementation preference I think (also assessing against all contracts might be kind of expensive computationally)

Reply:
yeah i think we can keep it simple and do the former (i.e. the current impl). The latter would require charging all contracts when the param is updated unless we want to store a lot of extra state, which isn't ideal